Update non-local-defns regarding parameterized traits and types (#3581)
rust-lang/rfcs: text/3373-avoid-nonlocal-definitions-in-fns.md

not cross-reference those definitions at all.

Humans cross-referencing such uses and definitions may find themselves
similarly baffled by code such as the following:
```rust
trait Trait<T> {
    fn method(&self) {}
}

struct Foo;

fn _foo() {
    struct Bar;
    impl Trait<Bar> for Foo {}
}

fn main() {
    Foo.method();
}

> ```

Reviewer: I'd prefer to move this example out of the motivation section.

# Explanation
[explanation]: #explanation

> An "expression-containing item" is defined as any [expression] where an [item]
> may be defined, for example:

Reviewer: ```suggestion
An "expression-containing item" is defined as any
[expression](https://doc.rust-lang.org/reference/expressions.html) where an
[item](https://doc.rust-lang.org/reference/items.html) may be defined. For
example:
```

# Explanation
[explanation]: #explanation

An "expression-containing item" is defined as any
[expression](https://doc.rust-lang.org/reference/expressions.html) where an
[item](https://doc.rust-lang.org/reference/items.html) may be defined. For
example:
- Functions
- Closures
- The values assigned to `static` items or non-anonymous `const` items.
- The discriminant values assigned to `enum` variants

> As a special exception, anonymous `const` items are excluded from this
> definition for the purpose of this RFC (see [unresolved-questions]).

Reviewer: ```suggestion
As an exception, anonymous `const` items are excluded from this
definition for the purposes of this RFC (see [unresolved-questions]),
as they're currently widely used to define items within macros.
```

# Unresolved questions
[unresolved-questions]: #unresolved-questions

Should we flag these definitions in anonymous `const` items as well? This is

> used in some macro expansions for [compatibility reasons][1].

Reviewer: ```suggestion
used in some macro expansions for
[compatibility reasons](https://github.com/rust-lang/rfcs/pull/3373#issuecomment-1885307786).
```

Is the last rule regarding parameterized trait `impl` items viable to implement?

# Future possibilities
[future-possibilities]: #future-possibilities

If in the future Rust provides a "standalone `derive`" mechanism (e.g. `derive
Trait for Type` as a standalone definition separate from `Type`), the `impl`
produced by that mechanism would be subject to the same requirements.

> 
> [expression]: https://doc.rust-lang.org/reference/expressions.html
> [item]: https://doc.rust-lang.org/reference/items.html
> [1]: https://github.com/rust-lang/rfcs/pull/3373#issuecomment-1885307786

Reviewer: ```suggestion
```
